src/metaprob/state.clj

```diff
@@ -35,15 +35,14 @@
     (cond (seq? state) (first state)
           (vector? state) (assert false "no value")
           (map? state)
-          (let [value (get state :value :no-value)]
-            (assert (not (= value :no-value)) ["state has no value" state])
-            value)
-          true (assert false ["not a state" state]))))
+          (do (assert (contains? state :value) ["state has no value" state])
+              (get state :value))
+          :else (assert false ["not a state" state]))))
 
 (defn has-subtrace? [state key]
   (if steady?
     (contains? (state-to-map state) key)
-    (cond (seq? state) (and (not (empty? state))
+    (cond (seq? state) (and (seq state)
```

Comment:
`seq` will allocate memory if state is `seqable?` but not a seq.  This is not good. I am not keen on this idiom.

Reply:
`empty?` is actually implemented in terms of `seq`: Here's [the source](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core.clj#L6126...L6131).